setup.ps1

```diff
@@ -0,0 +1,229 @@
+      Write-Log "Invalid input: $selection. Please enter letters between A and Z." -Level 'ERROR'
+    }
+  }
+  
+  return $indices
+}
+
+function Exit-Script {
+  param (
+    [int]$exitCode,
+    [bool]$isSubShell = $true,
+    [bool]$waitForKeypress = $true
+  )
+
+  if ($OldVirtualPath) {
+    $env:PATH = $OldVirtualPath
+  }
+
+  if ($exitCode -ne 0 -and $isSubShell) {
+    Write-Log "Script failed. Would you like to open the log file? (Y/N)" -Level 'PROMPT'
+    $openLog = Read-Host
+    if ($openLog -eq 'Y' -or $openLog -eq 'y') {
+      Start-Process notepad.exe -ArgumentList $LogFilePath
+    }
+  }
+
+  if ($waitForKeypress) {
+    Write-Log "Press any key to exit..."
+    $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown") | Out-Null
+  }
+
+  return $exitCode
+}
+
+# Function to handle installation
+function Install {
+  param ([string]$InputPath)
+  
+  if (-not $InputPath) {
+    Write-Log "No input provided for installation." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+
+  Write-Log "Installing $InputPath..."
+  $installCmd = if (Test-Path $InputPath) { $VenvPip + @('install', '-r', $InputPath) } else { $VenvPip + @('install', $InputPath) }
+  $output = & $installCmd[0] $installCmd[1..$installCmd.Length] 2>&1
+  $output | Out-File $LogFilePath -Append
+  if ($LASTEXITCODE -ne 0) {
+    Write-Log "Conflict detected. Exiting now..." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+}
+
+# Function to get the installed Python version tag for wheel compatibility
+function Get-PythonVersionTag {
+  $pythonVersion = & python -c "import sys; print(f'cp{sys.version_info.major}{sys.version_info.minor}')"
+  $architecture = & python -c "import platform; print('win_amd64' if platform.machine().endswith('64') else 'win32')"
+  return "$pythonVersion-$architecture"
+}
+
+# Exit in test environment
+if ($MyInvocation.InvocationName -ne $MyInvocation.MyCommand.Name) {
+  exit
+}
+
+# Check for admin privileges and elevate if necessary
+if (-NOT ([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
+  Write-Log "Requesting administrative privileges..." -Level 'ERROR'
+  Start-Process PowerShell -ArgumentList "-File `"$PSCommandPath`"" -Verb RunAs
+  Exit-Script -exitCode 1 -isSubShell $false
+}
+
+# Log file setup
+"Admin rights confirmed" | Out-File $LogFilePath -Append
+"Starting the script operations..." | Out-File $LogFilePath -Append
+
+# Navigate to the project directory
+Set-Location -Path $ProjectDir
+"Current directory: $(Get-Location)" | Out-File $LogFilePath -Append
+
+# Define the path to the Python executable in the virtual environment
+$VenvPython = Join-Path $ProjectDir "$VenvName\Scripts\python.exe"
+
+# Check if the virtual environment exists, if not, create it
+if (-Not (Test-Path $VenvPython)) {
+  Write-Log "Virtual environment not found. Creating virtual environment..." -IsError $false
+  python -m venv "$VenvName"
```

Comment:
How will "python not available" be handled? (especially considering that newer versions of python open a popup for this with an - in my experience odd - exit code)?
I assume the code will continue - and fail as python didn't actually create a virtual environment ?

Reply:
I will look into that